user_sync/rules.py

```diff
@@ -452,6 +467,32 @@ def sync_umapi_users(self, umapi_connectors):
                         self.updated_user_keys.add(user_key)
                     self.create_umapi_user(user_key, groups_to_add, umapi_info, umapi_connector)
 
+    def create_umapi_groups(self, umapi_connectors):
+        """
+        This is where we create user-groups. If auto_create is enabled,
+        this will pull user-groups from console and compare with mapped_groups. If mapped group does exist
+        in the console, then it will create. Note: Push Mode is not supported
+        :type umapi_connectors: UmapiConnectors
+        """
+        if not self.push_umapi:
+            umapi_info, umapi_connector = self.get_umapi_info(
+                PRIMARY_UMAPI_NAME), umapi_connectors.get_primary_connector()
+            mapped_groups = umapi_info.get_non_normalize_mapped_groups()
```

Comment:
OK, somehow this comment got lost from my review.

I think what you are trying to do with respecting LDAP group case and "non-normalized" groups is going to get clients into trouble, and it exacerbates a possible case that you should be checking for currently but aren't.

Because of the fact that you're doing case-sensitive matching, but more generally because of the way search/replace works, it's possible for more than one source group to map, via the same or different rules, to the same target group, and it's also possible for there to be two target groups that differ only in case.  That's going to be really bad for clients, who almost certainly don't want that to happen.

So I think you need to build logic in this function that catches this issue, so that if two different source groups map into the same target group you throw an error.  (If it turns out that a customer really wants to do this, then they can accomplish it by using `|` patterns in their source expressions.)  And I think that target-group mapping needs to be case-insensitive (actually normalized), so that you are always mapping source groups to their normalized group name, and the collision detection happens after normalization.

Finally, as noted elsewhere, I think you should annotate each target group with the name of the source group it was created from.  And I think you should really create each group in normalized form, but of course if you feel the target case is important you can create them as they were mapped.  But don't keep track of them that way, just keep track with each group of the case to be used to create it.

All of this normalization and error checking should happen *after* all the source groups are read but before any of the creations happen.

Reply:
Thanks - I will implement normalization and collision checking.  I'm planning on refactoring additional group resolution to the directory connector (possibly to the parent class) so I'll probably wait until after that is done before adding this.

Reply:
@adobeDan To implement collision checking, I created a data structure to keep track of the mapped additional target groups and the source groups from which they were derived.  If any target groups were derived from more than one source group then it throws an exception.

With this approach, I don't think it is possible to get around it with an or `|` pattern in the regular expression.  I can't think of any alternative approach that would allow it and still do accurate collision checking.

I don't think that'd a problem, but I wanted to make you aware.